### PR TITLE
Group 7 submission: 2352906 2352805 2352897 2352479 2350016 

### PR DESCRIPTION
**Pull request by Windyy-1007**
# Team members
1. Ôn Đức Phát - 2352897
2. Nguyễn Hữu Nam Phong - 2352906
3. Nguyễn Hoàng Khang - 2352479
4. Huỳnh Kim Nghĩa - 2352805
5. Phan Tấn Lộc - 2350016

# Algorithms
## Optimized First Fit Algorithm
The 2D stock cutting problem involves placing smaller rectangular pieces (products) onto larger rectangular sheets (stocks) while minimizing waste. The First Fit algorithm is a heuristic, greedy approach that attempts to place each piece in the first available position where it fits. The algorithm sort the products and stocks before processing. This approach return trim-loss of 0.15 to 0.2, with additional ~30% fill_ratio (number of stocks) saved.

## Params-allowed Genetic Algorithm
The Params-allowed genetic algorithm approach solves the 2D stock cutting problem through evolutionary optimization, maintaining a population of candidate solutions that evolves over generations to find optimal placements, with an addition of perks such as allow_sticky which grant bonus reward for products that stick to the side of the stock or another product, sort product, made a configed first move or focus on non-empty stocks.

For this submission, we suggest two settings:
### Setting 1 - Showcase (Run in ~25mins with fair optimization of trim loss = 0.3 to 0.45)
self, policy_id=2,
population_size=300,
mutation_rate=0.05,
generations=100,
stock_focus_weight=0.90,
allow_sticky = True,
sorted_products = False,
elitism_rate=0.10,
pre_firstmove = False
### Setting 2 - Aim to optimal (Run in ~2 hours with good optimization of trim loss = 0.12 to 0.3)
self, policy_id=2,
population_size=1000,
mutation_rate=0.05,
generations=200,
stock_focus_weight=0.90,
allow_sticky = True,
sorted_products = False,
elitism_rate=0.10,
pre_firstmove = False

### CODE REVIEW

**Comment by martinakaduc** [comment]
Please revise the PR in Dec 16. Otherwise, your group will get 0 score for code.

**Comment by Windyy-1007** [comment]
We have revise the PR. Can you check our PR again? Thank you and sorry for the inconvenience we have caused.
